### PR TITLE
feat: add semantic-release.

### PR DESCRIPTION
**Pull request by aorinevo**
## Changes

- Add semantic-release
- Updated `/docs/releasing`

## Notes
- addresses semantic-release part of https://github.com/NerdWalletOSS/shepherd/issues/111

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -63,6 +63,7 @@
     "simple-git": "^1.132.0"
   },
   "devDependencies": {
+    "@semantic-release/github": "^7.0.5",
```
Comment (aorinevo) [inline package.json:66]: Should be semantic-release.

**Comment by aorinevo** [comment]
@parshap, code changes are complete but we need to add `GH_TOKEN` and `NPM_TOKEN` secrets to this repo (see https://github.com/NerdWalletOSS/shepherd/settings/secrets).

Otherwise, this PR is ready for review.

**Comment by aorinevo** [comment]
Created and added `GH_TOKEN`.

**Comment by parshap** [comment]
@aorinevo what `GH_TOKEN` are you using? Also, won't we need an npm token as well for publishing?

**Comment by aorinevo** [comment]
I generated a GH_TOKEN off my account with proper repo perms.  

No need for NPM token as publishing is done via travis and only for new tags.  Github workflow will take care of CI/CD up to and including tagging.  Travis will take over from there and publish to npm.

File: README.md [hunk 2]
```diff
@@ -134,3 +134,13 @@ Shepherd is written in TypeScript, which requires compilation to JavaScript. Whe
...
 
 We use [TSLint](https://github.com/palantir/tslint) to ensure a consistent coding style and to help prevent certain classes of problems. Run `npm run lint` to run the linter, and `npm run fix-lint` to automatically fix applicable problems.
+
+### Semantic Releases
+
+Shepherd leverages [semantic-releases](https://github.com/semantic-release/semantic-release) to automate GitHub releases, including:
+- CHANGELOG updates made via [@semantic-release/changelog](https://github.com/semantic-release/changelog)
+- Release Notes compiled via [@semantic-release/release-notes-generator](https://github.com/semantic-release/release-notes-generator)
+- Tagging based on commit message via [@semantic-release/commit-analyzer](https://github.com/semantic-release/commit-analyzer)
+  - Commits that contains BREAKING CHANGE, BREAKING CHANGES, or BREAKING in their body will be considered breaking changes.
```
Comment (piperchester) [inline README.md:144]:
```suggestion
  - Commits that contain `BREAKING CHANGE`, `BREAKING CHANGES`, or `BREAKING` in their body will be considered breaking changes.
```

File: README.md [hunk 3]
```diff
@@ -134,3 +134,13 @@ Shepherd is written in TypeScript, which requires compilation to JavaScript. Whe
...
+  - Commits with a 'docs', 'style', 'refactor', or 'perf' type will be associated with a patch release
```
Comment (piperchester) [inline README.md:145]:
```suggestion
  - Commits with a `'docs'`, `'style'`, `'refactor'`, or `'perf'` type will be associated with a patch release.
```

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        with:
+          path: ~/.npm
+          key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-npm-
+      - name: Install dependencies
+        run: npm ci
+      - name: Release
+        env:
+          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
```
Comment (piperchester) [inline .github/workflows/release.yml:28]: Can you add a note to the "Semantic Releases" section in docs below if users need to add / maintain a GH token?

Comment (aorinevo) [reply .github/workflows/release.yml:28]: Yes sir.  Will do.

File: README.md [hunk 5]
```diff
@@ -134,3 +134,15 @@ Shepherd is written in TypeScript, which requires compilation to JavaScript. Whe
 Shepherd currently has minimal test coverage, but we're aiming to improve that with each new PR. Tests are written with Jest and should be named in a `*.test.ts` alongside the file under test. To run the test suite, run `npm run test`.
 
 We use [TSLint](https://github.com/palantir/tslint) to ensure a consistent coding style and to help prevent certain classes of problems. Run `npm run lint` to run the linter, and `npm run fix-lint` to automatically fix applicable problems.
+
+### Semantic Releases
```
Comment (parshap) [inline README.md:138]: Could you update (or remove) the release documentation here: https://github.com/NerdWalletOSS/shepherd/blob/master/docs/releasing.md

File: .travis.yml [hunk 6]
```diff
@@ -3,9 +3,6 @@ node_js:
   - 8
   - 10
   - 12
-cache:
-  directories:
-    - node_modules
```
Comment (parshap) [inline .travis.yml:8]: Why this change? I believe this is the default maybe?

File: README.md [hunk 7]
```diff
@@ -134,3 +134,15 @@ Shepherd is written in TypeScript, which requires compilation to JavaScript. Whe
...
+
+Shepherd leverages [semantic-release](https://github.com/semantic-release/semantic-release) to automate GitHub releases, including:
+- CHANGELOG updates made via [@semantic-release/changelog](https://github.com/semantic-release/changelog)
+- Release Notes compiled via [@semantic-release/release-notes-generator](https://github.com/semantic-release/release-notes-generator)
+- Tagging based on commit message via [@semantic-release/commit-analyzer](https://github.com/semantic-release/commit-analyzer)
+  - Commits that contain `BREAKING CHANGE`, `BREAKING CHANGES`, or `BREAKING` in their body will be considered breaking changes.
+  - Commits with a `docs`, `style`, `refactor`, or `perf` type will be associated with a patch release.
+  - If a commit doesn't match any of the above rules, it will be evaluated against the [default release rules](https://github.com/semantic-release/commit-analyzer/blob/master/lib/default-release-rules.js).
+
+  Note: `semantic-release` depends on `GH_TOKEN` secret being to a GitHub token with proper perms. Users need not set this locally.
```
Comment (parshap) [inline README.md:148]: I don't know if this relevant information, feel free to ignore if not.

```suggestion
Note: `semantic-release` depends on `GH_TOKEN` secret being to a GitHub token with proper perms. Users need not set this locally. The token was generated and added as a [secret](https://github.com/NerdWalletOSS/shepherd/settings/secrets) by @aorinevo.
```

Also, I didn't think this was meant to be nested within the above bullet, so I removed the initial spaces.

File: .releaserc.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                releaseRules: [
+                    ...typesForPatch.map((type) => ({
+                        type,
+                        release: 'patch'
+                    }))
+                ],
+                parserOpts: {
+                    noteKeywords: ['BREAKING CHANGE', 'BREAKING CHANGES', 'BREAKING']
+                }
+            }
```
Comment (parshap) [inline .releaserc.js:23]: Nit: A few brief comments explaining what these option overrides are accomplishing and *why* would be good.

File: .releaserc.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        ],
+        [
+            '@semantic-release/release-notes-generator',
+            {
+                parserOpts: {
+                    noteKeywords: ["BREAKING CHANGE", "BREAKING CHANGES", "BREAKING"]
```
Comment (parshap) [inline .releaserc.js:29]: Nit: Put this array in a `const` shared between the two places occurs.

**Comment by parshap** [comment]
Is there something that will "lint" commit messages to help contributors conform to the expectations of semantic-release? The semantic-release readme referenced commitizen or commitlint.

File: README.md (hunk 5 above)
Comment (aorinevo) [reply README.md:138]: Will do.

File: .releaserc.js (hunk 8 above)
Comment (aorinevo) [reply .releaserc.js:23]: I'd advocate to keep references to the source of truth, perhaps include links via comments?

Note, the links are included in `docs/releasing.md` doc.

File: .travis.yml (hunk 6 above)
Comment (aorinevo) [reply .travis.yml:8]: Caching node_modules directory had unexpected side effects on CI.  In particular, I got the following error:
```bash
The command "eval npm ci  " failed. Retrying, 2 of 3.
```

Removed cache as per: https://github.com/travis-ci/travis-ci/issues/10113#issuecomment-421121316

**Comment by aorinevo** [comment]
> Is there something that will "lint" commit messages to help contributors conform to the expectations of semantic-release? The semantic-release readme referenced commitizen or commitlint.

The [Semantic PR](https://github.com/zeke/semantic-pull-requests) Github app can help with this.

File: .releaserc.js (hunk 8 above)
Comment (parshap) [reply .releaserc.js]: Yep, links to the relevant documentation would be good. Even `See docs/releasing.md for details` if you think that's the most relevant.
